src/host/history.cpp

```diff
@@ -451,64 +431,33 @@ void CommandHistory::_Inc(SHORT& ind) const
     }
 }
 
-std::wstring CommandHistory::Remove(const SHORT iDel)
+std::wstring CommandHistory::Remove(const Index iDel)
 {
-    SHORT iFirst = 0;
-    auto iLast = gsl::narrow<SHORT>(_commands.size() - 1);
-    auto iDisp = LastDisplayed;
-
-    if (_commands.size() == 0)
+    if (iDel < 0 || iDel >= GetNumberOfCommands())
     {
         return {};
     }
 
-    const auto nDel = iDel;
-    if ((nDel < iFirst) || (nDel > iLast))
-    {
-        return {};
-    }
+    const auto str = std::move(_commands.at(iDel));
+    _commands.erase(_commands.begin() + iDel);
 
-    if (iDisp == iDel)
+    if (LastDisplayed == iDel)
     {
         LastDisplayed = -1;
     }
-
-    try
+    else if (LastDisplayed > iDel)
     {
-        const auto str = _commands.at(iDel);
-
-        if (iDel < iLast)
-        {
-            _commands.erase(_commands.cbegin() + iDel);
-            if ((iDisp > iDel) && (iDisp <= iLast))
-            {
-                _Dec(iDisp);
-            }
-            _Dec(iLast);
-        }
-        else if (iFirst <= iDel)
-        {
-            _commands.erase(_commands.cbegin() + iDel);
-            if ((iDisp >= iFirst) && (iDisp < iDel))
-            {
-                _Inc(iDisp);
-            }
-            _Inc(iFirst);
-        }
-
-        LastDisplayed = iDisp;
-        return str;
+        _Dec(LastDisplayed);

```

Comment:
We lost the increment/move to next command version (491-486 Left). When did it happen? Why is it OK to remove?

Reply:
I originally didn't intend to refactor this function, but it really puzzled me what it does. Hence the rewrite - it's now much more straightforward IMO.

I only understood it when I approached the function from a "first principles" kind of viewpoint, asking myself what a remove function would have to do:
* Ensure that the given index is in bounds
* Remove the given index
* Since all items after the given index have shifted one slow to the left, any index referring to those items also needs to shift one to the left

That last point is what your question is about: If the `LastDisplayed` value is after the given index, the item it was referring to has shifted to the left in the vector, and so `LastDisplayed` must be decremented.

The previous code was only that complex because the history used to be a ring buffer and there used to be not 1 but 3 indices that had to be shifted around. And it was possible that the items would shift to the right (due to it being a ring buffer) and so `LastDisplayed` had to be incremented.
Due to the switch to a `std::vector` none of that is necessary anymore.